rust_snuba/rust_arroyo/src/processing/strategies/reduce.rs

```diff
@@ -0,0 +1,202 @@
+use crate::processing::strategies::{CommitRequest, MessageRejected, ProcessingStrategy};
+use crate::types::{AnyMessage, InnerMessage, Message, Partition};
+use std::collections::BTreeMap;
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, SystemTime};
+
+struct BatchState<T, TResult> {
+    value: Option<TResult>,
+    accumulator: Arc<Mutex<dyn FnMut(TResult, T) -> TResult + Send + Sync>>,
+    offsets: BTreeMap<Partition, u64>,
+    batch_start_time: SystemTime,
+    message_count: usize,
+    is_complete: bool,
+}
+
+impl<T: Clone, TResult: Clone> BatchState<T, TResult> {
+    fn new(initial_value: TResult, accumulator: Arc<Mutex<dyn FnMut(TResult, T) -> TResult + Send + Sync>>) -> BatchState<T, TResult> {
+        BatchState {
+            value: Some(initial_value),
+            accumulator,
+            offsets: Default::default(),
+            batch_start_time: SystemTime::now(),
+            message_count: 0,
+            is_complete: false,
+        }
+    }
+
+    fn add(&mut self, message: Message<T>) {
+        let tmp = self.value.take();
+        self.value = Some((self.accumulator.lock().unwrap())(tmp.unwrap(), message.payload()));
```

Comment:
@untitaker Changed to `Arc<Mutex<dyn FnMut>>`... should I be handling the poisoned mutex error here?

Reply:
oh. uhh, please change `FnMut` to `Fn` to avoid the Mutex. Alternatively please try Rc instead of Arc.

Reply:
Ah, thanks. Removed the FnMut/Mutex